### PR TITLE
Fat 5550 create validatesandfilters

### PR DESCRIPTION
**Pull request by Activict**

### CODE REVIEW

File: StoreApp.BLL/Services/BrandService.cs [hunk 1]
```diff
@@ -58,6 +59,18 @@ public IEnumerable<BrandDTO> GetAll()
             return config.Map<IEnumerable<Brand>, List<BrandDTO>>(DataBase.Brands.GetAll());
         }
 
+        public int? CountProductsInCategory(int id)
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/BrandService.cs]: Look at the name of this method. I suppose to change this name)

File: StoreApp.BLL/Services/BrandService.cs [hunk 2]
```diff
@@ -58,6 +59,18 @@ public IEnumerable<BrandDTO> GetAll()
...
+        {
+            var count = DataBase.Products.GetAll().Count(p => p.BrandId.Equals(id));
+
+            if (count == 0)
+            {
+                return null;
+            }
+
+            return count;
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/BrandService.cs]: It seems to me as code which can be optimized

File: StoreApp.BLL/Services/CategoryService.cs [hunk 3]
```diff
@@ -63,6 +63,18 @@ public IEnumerable<CategoryDTO> GetAll()
...
+        public int? CountProductsInCategory(int id)
+        {
+            int count = DataBase.Products.GetAll().Count(p => p.CategoryId.Equals(id));
+            
+            if (count == 0)
+            {
+                return null;
+            }
+
+            return count;
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/CategoryService.cs]: The same as previous

File: StoreApp.BLL/Services/ProducerService.cs [hunk 4]
```diff
@@ -64,6 +65,18 @@ public IEnumerable<ProducerDTO> GetAll()
             return config.Map<IEnumerable<Producer>, IEnumerable<ProducerDTO>>(DataBase.Producers.GetAll());
         }
 
+        public int? CountProductsInCategory(int id)
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/ProducerService.cs]: The same as the previous note

File: StoreApp.BLL/Services/ProducerService.cs [hunk 5]
```diff
@@ -64,6 +65,18 @@ public IEnumerable<ProducerDTO> GetAll()
...
+        {
+            var count = DataBase.Products.GetAll().Count(p => p.ProducerId.Equals(id));
+
+            if (count == 0)
+            {
+                return null;
+            }
+
+            return count;
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/ProducerService.cs]: The same as the previous note

File: StoreApp/Controllers/BrandsController.cs [hunk 6]
```diff
@@ -110,6 +110,8 @@ public ActionResult DeleteBrand(int id)
                 return RedirectToAction("Index");
             }
 
+            TempData["countProducts"] = brandService.CountProductsInCategory(id);
```
Comment (AleksandroBoko) [inline StoreApp/Controllers/BrandsController.cs]: Need to be changed

File: StoreApp/Controllers/ProducersController.cs [hunk 7]
```diff
@@ -108,6 +108,8 @@ public ActionResult DeleteProducer(int id)
                 return RedirectToAction("Index");
             }
 
+            TempData["countProducts"] = producerService.CountProductsInCategory(id);
```
Comment (AleksandroBoko) [inline StoreApp/Controllers/ProducersController.cs]: Need to be changed

File: StoreApp/Views/Producers/DeleteProducer.cshtml [hunk 8]
```diff
@@ -10,24 +10,34 @@
...
                 <input type="submit" value="Delete" class="btn btn-danger" />
             </div>
         </div>
-    </div>
+    }
+    else
+    {
+        <div class="form-group">
+            <div class="col-md-offset-2 col-md-10 btn-danger">
+                If you want delete @Model.Name, you must delete @TempData["countProducts"] product(s)
```
Comment (AleksandroBoko) [inline StoreApp/Views/Producers/DeleteProducer.cshtml]: Some grammar mistake "If you want **to** delete"

File: StoreApp/Views/Category/DeleteCategory.cshtml [hunk 9]
```diff
@@ -6,27 +6,35 @@
...
+                    <input type="submit" value="Delete" class="btn btn-danger" />
+                </div>
             </div>
-        </div>
+        }
+        else
+        {
+            <div class="form-group">
+                <div class="col-md-offset-2 col-md-10 btn-danger">
+                    If you want delete @Model.Name, you must delete @TempData["countProducts"].ToString() product(s)
```
Comment (AleksandroBoko) [inline StoreApp/Views/Category/DeleteCategory.cshtml]: Some grammar mistake "If you want to delete"

File: StoreApp/Views/Brands/DeleteBrand.cshtml [hunk 10]
```diff
@@ -15,15 +15,26 @@
...
+                    <input type="submit" value="Delete" class="btn btn-danger" />
+                </div>
             </div>
-        </div>
+        }
+        else
+        {
+            <div class="form-group">
+                <div class="col-md-offset-2 col-md-10 btn-danger">
+                    If you want delete @Model.Name, you must delete @TempData["countProducts"] product(s)
```
Comment (AleksandroBoko) [inline StoreApp/Views/Brands/DeleteBrand.cshtml]: Some grammar mistake "If you want to delete"

File: StoreApp/Models/Store/CreateProductViewModel.cs [hunk 11]
```diff
@@ -9,8 +9,11 @@ public class CreateProductViewModel
         [Required]
         public string Name { get; set; }
         [Required]
+        [RegularExpression(@"^\d*(,)?(\d?\d?)?$", ErrorMessage = "Invalid characters")]
+        [Range(0.01, 999999999999, ErrorMessage = "Invalid value")]
         public decimal Price { get; set; }
```
Comment (AleksandroBoko) [inline StoreApp/Models/Store/CreateProductViewModel.cs:14]: You have set up the range of values here but you check the value before (**productDTO.Price.Equals(0)**)

File: StoreApp/Controllers/StoreController.cs [hunk 12]
```diff
@@ -94,13 +109,30 @@ public ActionResult EditProduct(EditProductViewModel product)
...
+                
                 return View(product);
             }
-            
-            productService.Edit(webMapper.config.Map<EditProductViewModel, ProductDTO>(product));
+
+            if (productDTO.Price.Equals(0))
+            {
+                productDTO.Enable = false;
+            }
```
Comment (AleksandroBoko) [inline StoreApp/Controllers/StoreController.cs]: Need to check this condition

File: StoreApp/Controllers/StoreController.cs [hunk 13]
```diff
@@ -52,16 +52,31 @@ public ActionResult CreateProduct()
...
                 TempData["Brands"] = new SelectList(brandService.GetAll(), dataValueField: "Id", dataTextField: "Name");
                 TempData["Producers"] = new SelectList(producerService.GetAll(), dataValueField: "Id", dataTextField: "Name");
                 return View(product);
             }
 
-            ProductDTO productDTO = webMapper.config.Map<CreateProductViewModel, ProductDTO>(product);
+            if (productDTO.Price.Equals(0))
+            {
+                productDTO.Enable = false;
+            }
```
Comment (AleksandroBoko) [inline StoreApp/Controllers/StoreController.cs]: Need to check this condition

File: StoreApp.BLL/Services/ProductService.cs [hunk 14]
```diff
@@ -58,6 +65,49 @@ public IEnumerable<ProductDTO> GetAll()
...
+                                                   p.ProducerId.Equals(productDTO.ProducerId));
+
+            products.ToList().ForEach(p => DataBase.Products.Detach(p));
+
+            if (product == null)
+            {
+                return true;
+            }
+
+            return false;
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/ProductService.cs]: You can apply a simpler code than this one

File: StoreApp.BLL/Services/ProductService.cs [hunk 15]
```diff
@@ -58,6 +65,49 @@ public IEnumerable<ProductDTO> GetAll()
             return config.Map<IEnumerable<Product>, IEnumerable<ProductDTO>>(DataBase.Products.GetAll());
         }
 
+        public bool ValidateNewProduct(ProductDTO product)
+        {
+            var validate = DataBase.Products.Find(p => p.Name.Equals(product.Name) &&
+                                                       p.BrandId.Equals(product.BrandId) &&
+                                                       p.ProducerId.Equals(product.ProducerId));
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/ProductService.cs]: It is not the correct name of the variable cause you get a collection of the objects here. You can name it like **products**

File: StoreApp.BLL/Services/ProductService.cs [hunk 16]
```diff
@@ -58,6 +65,49 @@ public IEnumerable<ProductDTO> GetAll()
...
+
+            foreach (var item in validate)
+            {
+                return false;
+            }
+
+            return true;
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/ProductService.cs]: I don't understand this approach. Need to discuss it

File: StoreApp.BLL/Services/ProductService.cs [hunk 17]
```diff
@@ -58,6 +65,49 @@ public IEnumerable<ProductDTO> GetAll()
...
+            var product = products.FirstOrDefault(p => p.Equals(productDTO.Id));
+
+            if (product != null &&
+                product.Name == productDTO.Name &&
+                product.BrandId == productDTO.BrandId &&
+                product.ProducerId == productDTO.ProducerId)
+            {
+                products.ToList().ForEach(p => DataBase.Products.Detach(p));
+                return true;
+            }
```
Comment (AleksandroBoko) [inline StoreApp.BLL/Services/ProductService.cs]: Probably, this block of code should be changed to optimize it.
Also we can divide this method into two parts. 
The first one is to check the equality of fields. The second one is to check if the product is existed by the these fields.

File: StoreApp.BLL/Services/BrandService.cs (hunk 2 above)
Comment (Activict) [reply StoreApp.BLL/Services/BrandService.cs]: fixed

File: StoreApp.BLL/Services/CategoryService.cs (hunk 3 above)
Comment (Activict) [reply StoreApp.BLL/Services/CategoryService.cs]: fixed

File: StoreApp.BLL/Services/ProducerService.cs (hunk 5 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProducerService.cs]: fixed

File: StoreApp/Views/Producers/DeleteProducer.cshtml (hunk 8 above)
Comment (Activict) [reply StoreApp/Views/Producers/DeleteProducer.cshtml]: fixed

File: StoreApp/Views/Category/DeleteCategory.cshtml (hunk 9 above)
Comment (Activict) [reply StoreApp/Views/Category/DeleteCategory.cshtml]: fixed

File: StoreApp/Views/Brands/DeleteBrand.cshtml (hunk 10 above)
Comment (Activict) [reply StoreApp/Views/Brands/DeleteBrand.cshtml]: fixed

File: StoreApp/Models/Store/CreateProductViewModel.cs (hunk 11 above)
Comment (Activict) [reply StoreApp/Models/Store/CreateProductViewModel.cs:14]: change to Quantity

File: StoreApp/Controllers/StoreController.cs (hunk 12 above)
Comment (Activict) [reply StoreApp/Controllers/StoreController.cs]: change to Quantity

File: StoreApp/Controllers/StoreController.cs (hunk 13 above)
Comment (Activict) [reply StoreApp/Controllers/StoreController.cs]: change to Quantity

File: StoreApp.BLL/Services/ProductService.cs (hunk 14 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProductService.cs]: fixed

File: StoreApp.BLL/Services/ProductService.cs (hunk 15 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProductService.cs]: fixed

File: StoreApp.BLL/Services/ProductService.cs (hunk 16 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProductService.cs]: fixed

File: StoreApp.BLL/Services/ProductService.cs (hunk 14 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProductService.cs]: fixed

File: StoreApp.BLL/Services/BrandService.cs (hunk 1 above)
Comment (Activict) [reply StoreApp.BLL/Services/BrandService.cs]: fixed

File: StoreApp.BLL/Services/ProducerService.cs (hunk 4 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProducerService.cs]: fixed

File: StoreApp/Controllers/BrandsController.cs (hunk 6 above)
Comment (Activict) [reply StoreApp/Controllers/BrandsController.cs]: fixed

File: StoreApp/Controllers/ProducersController.cs (hunk 7 above)
Comment (Activict) [reply StoreApp/Controllers/ProducersController.cs]: fixed

File: StoreApp.BLL/Services/ProductService.cs (hunk 17 above)
Comment (Activict) [reply StoreApp.BLL/Services/ProductService.cs]: fixed

**Pull request merged**